Snapshot description

Laborratte5 opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------
This PR adds the ability to add a description text to snapshots.
When creating a new snapshots with `restic backup` the `--description`
and `--description-file` option allows setting a description text.

The description of existing snapshots can be changed using the
`restic description <snapshotID> [--set <desc> | --set-file <file>]` subcommand.
To remove a description `restic description <snapshotID> --remove-description` can be used.

<!--
Describe the changes and their purpose here, as detailed as needed.
-->

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->
https://github.com/restic/restic/issues/2376

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

- [X] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [X] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I'm done! This pull request is ready for review.

Laborratte5 committed:
test backup description option

Laborratte5 committed:
Store snapshot description

Laborratte5 committed:
Show snapshot description when listing snapshots

Laborratte5 committed:
test description command

Test if the description command changes the description of
existing snapshots by creating a new snapshot with altered
description.

Laborratte5 committed:
Implement runDescription to change the description for a snapshot with the given ID

Laborratte5 committed:
refactor: Group options for decription command

Group options for the description command
in DescriptionOptions struct

Laborratte5 committed:
test changing and removing snapshot description

Laborratte5 committed:
refactor: remove duplication of description texts

Laborratte5 committed:
description command: use positional arguments instead of flags

Use positional arguments for the description text and snapshot list
instead of command flags

Laborratte5 committed:
Add a description command to change the description of existing snapshots

Laborratte5 committed:
description: do not create new snapshot when description did not change

Laborratte5 committed:
Add changelog file

Laborratte5 commented:
Note: I intentionally didn't add support for changing the description of multiple snapshots add once
as I think descriptions a somewhat unique per snapshot. This prevents users from accidentally overwriting
the description of many snapshots.

Laborratte5 commented:
Another idea that just came to my mind is the following.
As multi-line descriptions look kind of weird in the snapshot table, maybe some formatting like git commit messages
could be used, where the first line is the heading, displayed in the snapshots table. And after a free line the longer
description would follow, which is only displayed with the `description` command.

The usage of the description command would then change to `restic description [snapshotID ...]` to show the complete description for all listed snapshots. (I would then also add the snapshot filter options like `--tag`, `--host` etc.)
To edit the description a `--edit` option would be added to the description subcommand.

```sh
Laborratte5@ARCH-MEDION ~/P/P/restic > ./restic snapshots --repo test-repo
ID        Time                 Host         Tags        Paths                                     Size         Description
--------------------------------------------------------------------------------------------------------------------------------------------
658d9f2e  2025-04-23 12:25:06  ARCH-MEDION              /home/Laborratte5/Projekte/Programmieren/restic  256.546 MiB  this is the first description
7d534f72  2025-04-23 13:31:35  ARCH-MEDION              /home/Laborratte5/Projekte/Programmieren/restic  532.816 MiB  my description
41a40da2  2025-04-23 13:32:21  ARCH-MEDION              /home/Laborratte5/Projekte/Programmieren/restic  695.289 MiB  this is a third description
afcaed29  2025-04-25 23:16:25  ARCH-MEDION              /home/Laborratte5/Projekte/Programmieren/restic  888.552 MiB  this is the first description
--------------------------------------------------------------------------------------------------------------------------------------------

Laborratte5@ARCH-MEDION ~/P/P/restic > ./restic description 41a40da2  --repo test-repo
this is a third description

longer description of the purpose and rational behind the snapshot.
lorem ipsum
yada yada
Laborratte5@ARCH-MEDION ~/P/P/restic > 
```

However in this case the description should be read via a file since multi line input with a cli may be a bad user experience.

Wonderer0 commented:
> However in this case the description should be read via a file since multi line input with a cli may be a bad user experience.

I definitely agree.  It would be a good idea to allow a `--description-file` option to the `backup` command, with the first line shown in the list of snapshots, similar to Git as you describe above.  This allows users to write the descriptions using their favourite text editor, and then make the backup using the same command every time.

If the user always puts the description file in the same place near the top of the directory being packed up it would be possible to write a script to extract the full details the backups and build a backup history text file from them. 

I personally would prefer a shorter name like `--notes` or `--comment` (like Borg), but that's a very minor issue.

Laborratte5 committed:
backup add description-file flag

Add description-file flag which can be used
to set the snapshot description to the content
of the given file.

Laborratte5 committed:
refactor: create extra function to read snapshot description input

Laborratte5 committed:
refactor snapshot descriptions

- bundle descriptionOptions in struct
- add explicit flag for removing snapshot descriptions

Laborratte5 committed:
description: only exclusivly lock repo when changing descriptions

Laborratte5 committed:
description: show complete snapshot description with description command

The complete description of a snapshot is shown when using
the `description` command without the `--description`
or `--description-file` flag.

Laborratte5 committed:
description: truncate description text in snapshot table

Laborratte5 committed:
update unreleased changelog

Laborratte5 committed:
refactor: move readDescription function to cmd_description

Laborratte5 committed:
Output information about changed snapshots when description is modified

Laborratte5 committed:
doc: add section about snapshot descriptions

aawsome commented:
Having this in the restic design document would be nice!
About the JSON name: Please keep `description` - this is the name rustic also already uses!

Laborratte5 commented:
> Having this in the restic design document would be nice!

I forgot about that. I'll add this to the doc/design.rst file under Snapshots.

> About the JSON name: Please keep description

Do you mean as the `message_type` of the json output?

aawsome commented:
Yes it is already the correct name. I just wanted to emphasize that this should not be changed! ;-)

Laborratte5 committed:
doc: Add snapshot description field to design reference docs

Laborratte5 commented:
I now added this to the design document. Let me know if this is good enough or if I should add more info. (Although I am unsure what else to add)

MichaelEischer commented on internal/ui/table/table.go:
```diff
@@ -25,6 +25,24 @@ type Table struct {
+	headline := strings.Split(s, "\n")[0]
+
+	if len(headline) > maxLen {
+		if len(headline) >= 3 {
```
```suggestion
		if len(headline) > 3 {
```

MichaelEischer commented on internal/ui/table/table.go:
```diff
@@ -25,6 +25,24 @@ type Table struct {
+			headline = headline[:maxLen]
+		}
+
+		headline += "..."
```
Shouldn't the `...` only be appended when the headline is shortened to `maxLen-3`? That way the resulting string actually has the requested length.

MichaelEischer commented on cmd/restic/cmd_snapshots.go:
```diff
@@ -256,6 +266,10 @@ func PrintSnapshots(stdout io.Writer, list restic.Snapshots, reasons []restic.Ke
 			data.Size = ui.FormatBytes(sn.Summary.TotalBytesProcessed)
 		}
 
+		if sn.Description != "" {
+			data.Description = sn.Description
+		}
```
```suggestion
		data.Description = sn.Description
```
The if is redundant.

MichaelEischer commented on cmd/restic/cmd_backup.go:
```diff
@@ -301,6 +303,8 @@ func (opts BackupOptions) Check(gopts GlobalOptions, args []string) error {
 		}
 	}
 
+	opts.DescriptionOptions.Check()
+
 	return nil
```
```suggestion
	return opts.DescriptionOptions.Check()
```

MichaelEischer commented on cmd/restic/cmd_description.go:
```diff
@@ -0,0 +1,216 @@
+		Long: `
+The "description" command allows you to view or modify the description on an existing snapshot.
+
+The special snapshotID "latest" can be used to restore the latest snapshot in the
```
Copy&paste typo

MichaelEischer commented on cmd/restic/cmd_description.go:
```diff
@@ -0,0 +1,216 @@
+	var opts changeDescriptionOptions
+
+	cmd := &cobra.Command{
+		Use:   "description snapshotID [--set description | --set-file description]",
```
Those option names don't match the naming in `descriptionOptions`.

MichaelEischer commented on cmd/restic/cmd_description.go:
```diff
@@ -0,0 +1,216 @@
+}
+
+func (opts *changeDescriptionOptions) Check() error {
+	err := opts.descriptionOptions.Check()
+	if err != nil {
+		return err
+	}
+	return nil
```
```suggestion
	return opts.descriptionOptions.Check()
```

MichaelEischer commented on cmd/restic/cmd_description.go:
```diff
@@ -0,0 +1,216 @@
+// readDescription returns the description text specified by either the
+// `--description` option or the content of the `--description-file`
+func readDescription(opts descriptionOptions) (string, error) {
+	descriptionScanner := bufio.NewScanner(strings.NewReader(opts.Description))
```
What is the bufio construction used for when `opts.Description != ""`? The function could just return the `opts.Description`.

MichaelEischer commented on cmd/restic/cmd_description.go:
```diff
@@ -0,0 +1,216 @@
+		}
+	}
+
+	if !(opts.removeDescription || descriptionChange) {
```
```suggestion
	else {
```

MichaelEischer commented on cmd/restic/cmd_description.go:
```diff
@@ -0,0 +1,216 @@
+				continue
+			}
+		}
+		return nil
```
```suggestion
```

MichaelEischer commented on cmd/restic/cmd_description.go:
```diff
@@ -0,0 +1,216 @@
+
+	if gopts.JSON {
+		printFunc = func(c changedSnapshot) {
+			term.Print(ui.ToJSONString(c))
```
The JSON output should also be documented at `doc/075_scripting.rst`.

Laborratte5 replied:
I removed the JSON output for now, since `rewrite` does not have JSON output currently.

MichaelEischer reviewed (changes requested):
I somewhat doubt that we need a `description` command. The `--set` and `--remove` options could just as well be added to the `rewrite` command (I'd expect rather little usage of those options). And the `snapshots` command should IMHO default to showing the full description (just like `git log` does by default). If `--compact` is passed, then it would be fine to truncate the description output.

What do you think?

The merge conflicts are most likely due to https://github.com/restic/restic/pull/5518 and https://github.com/restic/restic/pull/5525 . Just let me know me if you need help with rebasing the PR.

Laborratte5 committed:
refactor: Apply proposed changes

Laborratte5 committed:
Check description size after reading it from file or cli

The current limit 'maxDescriptionLength'
is set to 4096 bytes which was chosen arbitrarily.

Laborratte5 committed:
Allow description command with multiple snapshot ids

Allow changing the description of multiple snapshots at once.

Laborratte5 committed:
fix: testRunDescriptionRemove should check return for errors

Laborratte5 committed:
Show complete snapshot description by default

Show the complete multiline description when running
`restic snapshot` by default. Only truncate description
when `--compact` is passed.

Laborratte5 committed:
fix: error check of description flags and usage text

Laborratte5 committed:
Remove description command

Laborratte5 committed:
Add changeDescriptionOptions to cmd_rewrite

Laborratte5 committed:
Add readDescription to cmd_rewrite

Laborratte5 committed:
description: Add test for rewrite description

Laborratte5 committed:
Change snapshot descriptions with rewrite command.

Laborratte5 committed:
Update snapshot description changelog

Laborratte5 committed:
Update documentation to explain how to change snapshot descriptions

MichaelEischer commented on changelog/unreleased/issue-2376:
```diff
@@ -0,0 +1,16 @@
+Enhancement: Add per snapshot description
+
+Restic can now store user supplied descriptions for snapshots.
```
```suggestion
Restic can now store user-supplied descriptions for snapshots.
```

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -100,13 +110,66 @@ func (sma snapshotMetadataArgs) convert() (*snapshotMetadata, error) {
+// changeDescriptionOptions collects all options for description changing
+type changeDescriptionOptions struct {
+	descriptionOptions
+	removeDescription bool
```
As the description is just a string field in the snapshot, it would also be possible to just use `--description ""`. I'm somewhat undecided whether that would be simpler to use or not.

Laborratte5 replied:
I do like the idea of the `--description ""`. However, it seems like there is no way to avoid the default value with `pflag`. This means there is no way of distinguishing if the default value of the flag was used because the flag was not set by the user (in which case the description should not be changed) or if the user provided an empty value in order to remove the description.
If there is a solution to this I am happy to implement it since I do like the simpler interface. However, I have not found a good way to solve this issue.

The only solution that came to my mind would be using a random string as a default value and then comparing the flag value with this string, but this feels too hacky to me.

MichaelEischer replied:
It's possible to check whether a flag was explicitly specified on the command line by checking the flag itself. Search for `readConcurrencyFlag` for an example.

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -100,13 +110,66 @@ func (sma snapshotMetadataArgs) convert() (*snapshotMetadata, error) {
+
+const maxDescriptionLength = 4096
+
+var descriptionTooLargeErr = errors.New(fmt.Sprintf("The provided descriptions exceeds the maximum length of %d bytes.", maxDescriptionLength))
```
```suggestion
var descriptionTooLargeErr = fmt.Errorf("the provided description exceeds the maximum length of %d bytes", maxDescriptionLength)
```

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -186,11 +255,11 @@ func rewriteSnapshot(ctx context.Context, repo *repository.Repository, sn *data.
 
 func filterAndReplaceSnapshot(ctx context.Context, repo restic.Repository, sn *data.Snapshot,
-	filter rewriteFilterFunc, dryRun bool, forget bool, newMetadata *snapshotMetadata, addTag string, printer progress.Printer) (bool, error) {
+	filter rewriteFilterFunc, dryRun bool, forget bool, newMetadata *snapshotMetadata, description string, addTag string, printer progress.Printer) (bool, error) {
```
A new description should be passed as part of `newMetadata *snapshotMetadata`. The description must also be passed as string pointer. Otherwise, there's no way to distinguish between "keep the current description" and "remove description".

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -266,6 +339,9 @@ func filterAndReplaceSnapshot(ctx context.Context, repo restic.Repository, sn *d
 	}
 
+	printer.P("setting description to %s", description)
+	sn.Description = description
```
Should only be changed if a change was requested.

MichaelEischer commented on cmd/restic/cmd_backup_integration_test.go:
```diff
@@ -467,6 +467,41 @@ func TestIncrementalBackup(t *testing.T) {
+
+	rtest.Assert(t, newest.Description == "",
+		"expected no description, got %v", newest.Description)
+	parent := newest
```
I don't really see the point of setting a parent snapshot here.

MichaelEischer reviewed (changes requested):
Thanks for the changes so far! I've found a bunch of small issues and a larger bug in the `rewrite` command. The current code always overwrites the description, such that it would overwrite an existing description even if not requested by the user.

We could also use `--description ""` instead of `--remove-description`. Although, I'm not sure whether that would be simpler to use.

rawtaz commented:
> We could also use --description "" instead of --remove-description. Although, I'm not sure whether that would be simpler to use.

The first one is better IMO. Since the description field will be there anyway, it's much better to just set the description to be empty, than to add additional options to act on it.

Laborratte5 commented:
As you have noticed I haven't come around working on this PR lately. I fixed some of the mentions above before realizing you reassigned this issue. Do you still want me to push those changes or should I leave it to you now to avoid more overhead?

MichaelEischer commented:
@Laborratte5 the assignee status just used to track who takes care of the PR reviews and follow ups.

Laborratte5 commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -100,13 +110,66 @@ func (sma snapshotMetadataArgs) convert() (*snapshotMetadata, error) {
+		}
+		descriptionScanner := bufio.NewScanner(bytes.NewReader(data))
+		var builder strings.Builder
+		for descriptionScanner.Scan() {
```
`Scanner` tokenizes the input line by line. The loop ensures all lines of the file are read.

MichaelEischer replied:
Doesn't `description, _ = strings.CutSuffix(string(data), "\n")` do the same as this loop?

Laborratte5 committed:
test read snapshot description from file

Laborratte5 committed:
test do not set parent snapshot in TestBackupDescription

A parent snapshot is not needed when testing wether the
description of a snapshot is set correclty or not.

MichaelEischer commented:
Are you still interested in finishing the PR?

Laborratte5 commented:
Yes, I am. I was waiting for feedback regarding this (sorry for the communication problems):

> I do like the idea of the `--description ""`. However, it seems like there is no way to avoid the default value with `pflag`. This means there is no way of distinguishing if the default value of the flag was used because the flag was not set by the user (in which case the description should not be changed) or if the user provided an empty value in order to remove the description. If there is a solution to this I am happy to implement it since I do like the simpler interface. However, I have not found a good way to solve this issue.
> 
> The only solution that came to my mind would be using a random string as a default value and then comparing the flag value with this string, but this feels too hacky to me.

I did push my current state. The only thing that is missing is the direction regarding the `--remove-decription` flag and the documentation which I try to finish this week.

Laborratte5 committed:
Improve wording and code-style

Laborratte5 committed:
Use ui.Truncate to truncate description in snapshot table

Laborratte5 committed:
Add test reading backup description from file

Laborratte5 committed:
Store snapshot description as pointer in snapshotMetadata

This makes it possible to encode wether the description of
a snapshot should be removed or should be kept as is during
snapshot rewrite. This also allows to only create a new
(rewritten) snapshot if the description was actually changed.

Laborratte5 committed:
Use fmt.Errorf instead of errors.New

Laborratte5 committed:
Omit unnecessary call to testRunCheck

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -97,7 +129,59 @@ func (sma snapshotMetadataArgs) convert() (*snapshotMetadata, error) {
+
+// readDescription returns the description text specified by either the
+// `--description` option or the content of the `--description-file`
+func readDescription(opts descriptionOptions) (string, error) {
```
Should move to `cmd_backup` next to the definition `descriptionOptions`. Also change this to a method of `descriptionOptions`. If you want, you can also extract the `descriptionOptions` into a separate file like `description.go` or similar.

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -97,7 +129,59 @@ func (sma snapshotMetadataArgs) convert() (*snapshotMetadata, error) {
+}
+
+// changeDescriptionOptions collects all options for description changing
+type changeDescriptionOptions struct {
```
Without `--remove-description` it should be possible to remove this struct. From what I can see `cmd_backup` shouldn't have problems if we explicitly track whether description was specified.

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -285,15 +387,18 @@ func filterAndReplaceSnapshot(ctx context.Context, repo restic.Repository, sn *d
 func runRewrite(ctx context.Context, opts RewriteOptions, gopts global.Options, args []string, term ui.Terminal) error {
 	if !opts.SnapshotSummary && opts.ExcludePatternOptions.Empty() && opts.Metadata.empty() {
-		return errors.Fatal("Nothing to do: no excludes provided and no new metadata provided")
+		return errors.Fatal("Nothing to do: no excludes provided, no new metadata provided and no description manipulation flag provided")
```
```suggestion
		return errors.Fatal("Nothing to do: no excludes provided, no new metadata provided and no description modification flag provided")
```

MichaelEischer reviewed (changes requested):
> The only thing that is missing is the direction regarding the `--remove-decription` flag

I've completely missed your question for that flag. I've answered below https://github.com/restic/restic/pull/5375/changes#r3311748124 .

Laborratte5 committed:
fixup! Implement runDescription to change the description for a snapshot with the given ID
